### DISCUSSION TITLE
Guidance for writing customisations for tests that have many permutations

### DISCUSSION THREAD

**Discussion by argium** [AutoFixture/AutoFixture#1482] [Q&A] [answered]
Hello, 

First of all - I really like the potential of AutoFixture but I'm struggling to learn the most optimal way writing tests that cover many different permutations and I'm spending 2-10x longer writing the tests compared to using AutoMoq directly.

In this case, I'm writing unit tests for ASP.NET authorization policies and DataAnnotations validators. The common pattern here is to:
1. Define the input (int, string, Uri, class, etc..)
2. Define the configuration (IOptionsSnapshot)
3. Instantiate the class being tested and call its method
4. Verify output

Often there are many options of both 1 and 2 resulting in dozens of permutations and my tests have turned into:

```cs
[Theory]
[InlineAutoMoqData(typeof(InputACustomisation), typeof(ConfigACustomisation))]
[InlineAutoMoqData(typeof(InputBCustomisation), typeof(ConfigBCustomisation))]
[InlineAutoMoqData(typeof(InputCCustomisation), typeof(ConfigCCustomisation))]
void Function_ReturnsTrue_ForValidInput(SUT sut, Input input) 
{
  var result = sut.DoWork(input);
  Assert.True(result);
}

[Theory]
[InlineAutoMoqData(typeof(InputBCustomisation), typeof(ConfigACustomisation))]
[InlineAutoMoqData(typeof(InputCCustomisation), typeof(ConfigACustomisation))]
[InlineAutoMoqData(typeof(InputACustomisation), typeof(ConfigBCustomisation))]
[InlineAutoMoqData(typeof(InputCCustomisation), typeof(ConfigBCustomisation))]
[InlineAutoMoqData(typeof(InputACustomisation), typeof(ConfigCCustomisation))]
[InlineAutoMoqData(typeof(InputBCustomisation), typeof(ConfigCCustomisation))]
void Function_ReturnsFalse_ForValidInput(SUT sut, Input input) 
{
  var result = sut.DoWork(input);
  Assert.False(result);
}
```

The problems I see here are:
1. It's hard to read 
2. It's easy to make mistakes, miss permutations
3. Each variant needs its own class (which we keep in their own files) creating a large amount of autofixture code

(2) may be solvable with xunit DataAttribute's to create central lists of related customisations but only in some cases
(3) I don't see a way around this until I move the customisations from attributes into the test method (because I can't have a more flexible customisation that accepts constructor arguments)

Composite customisations don't appear to help here and only create *more* files to manage, some of which may only be relevant for this specific set of tests. For example:

```cs
// 4 files
[Theory]
[InlineAutoMoqData(typeof(InputAWithConfigACustomisation))]
[InlineAutoMoqData(typeof(InputBWithConfigBCustomisation))]
[InlineAutoMoqData(typeof(InputCWithConfigCCustomisation))]
void Function_ReturnsTrue_ForValidInput(SUT sut, Input input) 
{
  var result = sut.DoWork(input);
  Assert.True(result);
}

// 5 files
class InputAWithConfigACustomisation : CompositeCustomisation
{
  public InputAWithConfigACustomisation() : base (new InputACustomisation(), new ConfigACustomisation) {}
}
```

Although this is a small example, the problem becomes much worse once I begin to consider the larger system (an ASP.NET API) which includes many dependencies, subcomponents, configurations, object models, and inputs. If the above examples were for unit testing authorization policies, a higher order example may be writing a test which:

1. Loads the appsettings for a particular asp.net core environment
2. Customises the services in the dependency container, including auth hooks and configuration. This may include ensuring consistency between configuration and mock service clients, only returning the expected values if the URI from the config is correctly used.
3. Customises the identity that is return by the AuthN module
4. Customises the configuration of the authZ policy
5. Customises request and response objects, sometimes to again ensure consistency with appsettings or sometimes to disable AutoMoq on specific problematic fields

```cs
[Theory]
[InlineAutoMoqData(typeof(InputB), typeof(Module1ConfigA), typeof(Module2ConfigA), typeof(DependenciesCustomisation))]
void Function_ReturnsFalse_ForValidInput(SUT sut, Input input) 
{
  var result = sut.DoWork(input);
  Assert.False(result);
}
```

I feel as though I'm not grasping a key concept and the tests that I'm writing are more complex than not using AutoFixture. Do you have any guidance or advice that might help?  If I removed the requirement of using attributes, I think the tests *could* become cleaner with the use of customisations that accept constructor arguments but is including IFixture in the test method parameters an anti-pattern?

**Comment by aivascu** [comment, marked as answer]
Hi @argium,
It feels like you're trying to hide too many details of the test setup.  Could you provide a sample of an application you'd like to test and a setup of your customizations and data attributes?

> is including IFixture in the test method parameters an anti-pattern?

No, it is not an anti-pattern. For example, it can be usefull to include the fixture in your test parameters when you want to build a entity/DTO during the test, with some custom properties.

```cs
[Theory]
[MyCustomData]
public void SampleTest(MySut sut, IFixture fixture)
{
    // Arrange
    var myDtos = fixture.Build<MyDto>()
        .With(x => x.MyArgument, "MyValue")
        .CreateMany();

    // Act
    var actual = syt.DoStuff(myDtos);

   /* ... */
}
```

**Comment by tonyhallett** [comment]
For NUnit https://github.com/AutoFixture/AutoFixture/discussions/1559
